core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
     getPartition(topicPartition) match {
       case HostedPartition.Online(partition) => Some(partition)
+      case HostedPartition.Fenced(metadata) => Some(metadata.partition)
       case HostedPartition.None | HostedPartition.Offline => None
     }
   }
 
+  // not sure if any of the 9 invocations of the above nonOfflinePartition() should be diverted here
+  def onlinePartition(topicPartition: TopicPartition): Option[Partition] = {
+    getPartition(topicPartition) match {
+      case HostedPartition.Online(partition) => Some(partition)
+      case _ => None
+    }
+  }
+
+  def fencedPartition(topicPartition: TopicPartition): Option[Partition] = {
+    getPartition(topicPartition) match {
+      case HostedPartition.Fenced(metadata) => Some(metadata.partition)
+      case _ => None
+    }
+  }
+
   // An iterator over all non offline partitions. This is a weakly consistent iterator; a partition made offline after
   // the iterator has been constructed could still be returned by this iterator.
   private def nonOfflinePartitionsIterator: Iterator[Partition] = {
     allPartitions.values.iterator.flatMap {
       case HostedPartition.Online(partition) => Some(partition)
+      case HostedPartition.Fenced(metadata) => Some(metadata.partition)
       case HostedPartition.None | HostedPartition.Offline => None
     }
   }
 
+  // An iterator over all online partitions. This is a weakly consistent iterator; a partition made offline or fenced
+  // after the iterator has been constructed could still be returned by this iterator.
+  private def onlinePartitionsIterator: Iterator[Partition] = {
+    allPartitions.values.iterator.flatMap {
+      case HostedPartition.Online(partition) => Some(partition)
+      case _ => None
+    }
+  }
+
   private def offlinePartitionCount: Int = {
     allPartitions.values.iterator.count(_ == HostedPartition.Offline)
   }
 
-  def getPartitionOrException(topicPartition: TopicPartition): Partition = {
-    getPartitionOrError(topicPartition) match {
+  // An iterator over all fenced partitions. This is a weakly consistent iterator; a partition made off/online
+  // after the iterator has been constructed could still be returned by this iterator.
+  private def fencedPartitionsIterator: Iterator[FencedPartitionMetadata] = {
+    allPartitions.values.iterator.flatMap {
+      case HostedPartition.Fenced(metadata) => Some(metadata)
+      case _ => None
+    }
+  }
+
+  def getOnlinePartitionOrException(topicPartition: TopicPartition): Partition = {
+    getOnlinePartitionOrError(topicPartition) match {
+      case Left(Errors.KAFKA_STORAGE_ERROR) =>
+        throw new KafkaStorageException(s"Partition $topicPartition is in an offline log directory")
+
+      case Left(error) =>
+        throw error.exception(s"Error while fetching partition state for $topicPartition")
+
+      case Right(partition) => partition
+    }
+  }
+
+  def getNonOfflinePartitionOrException(topicPartition: TopicPartition): Partition = {
```

Comment:
Similar to other comment. Can we refuse to expose `Partition` while it is fenced?

Reply:
Are we going to allow follower fetching when the broker is fenced? if so, then I wonder if it becomes necessary to expose the Partition even when we have deferred changes.

Reply:
If the broker is fenced then we won't be getting any new metadata, and nothing will defer.  Everything related to deferred partitions is now scoped private within ReplicaManager.